working/0546-patterns/patterns-feature-specification.md

```diff
@@ -2159,17 +2209,233 @@ To match a pattern `p` against a value `v`:
+
+1.  Let `i` be an invocation key with no parent, no extension type, named
+    `this`, with an empty argument list. *This is the root node of the
+    invocation key tree and represents the matched value itself.*
+
+2.  For each pattern `p` in `s` with parent invocation `i`, bind invocation keys
+    to it and its subpatterns using the following procedure:
+
+To bind invocation keys in a pattern `p` using parent invocation `i`:
+
+*   **Logical-or** or **logical-and**:
+
+    1.  Bind invocations in the left and right subpatterns using target `i`.
+
+*   **Relational**:
+
+    1.  Resolve the relational operator to a member.
+
+    2.  If it resolves to an instance method, then bind `i : (op, [arg])` to
+        the operator method invocation where `op` is the name of the operator
+        and `arg` is the right operand value.
+
+    3.  Else it resolves to an extension method. Bind `i : type(op, [arg])` to
+        the operator method invocation where `type` is the type the extension
+        method resolved to, `op` is the name of the operator and `arg` is the
+        right operand value.
+
+*   **Cast**, **null-check**, **null-assert**, or **parenthesized**:
+
+    1.  Bind invocations in the subpattern using target `i`.
+
+*   **Constant**:
+
+    1.  Bind `i : ("==", [arg])` to the `==` method invocation where`arg` is the
+        right operand value.
+
+*   **Variable**:
+
+    1.  Nothing to do.
+
+*   **List**:
+
+    1.  Bind `i : ("length", [])` to the `length` getter invocation.
+
+    2.  For each element subpattern:
+
+        1.  Let `e` be `i : ("[]", [index])` where `index` is the zero-based
+            index of this element subpattern.
+
+        2.  Bind `e` to the `[]` invocation for this element.
+
+        3.  Bind invocations in the element subpattern using target `e`.
+
+*   **Map**:
+
+    1.  Bind `i : ("length", [])` to the `length` getter invocation.
+
+    2.  For each entry in `p`:
+
+        1.  Bind `i : ("containsKey()", [key])` to the `containsKey()`
+            invocation where `key` is entry's key constant value.
+
+        2.  Let `e` be `i : ("[]", [key])` where `key` is entry's key constant
+            value.
+
+        3.  Bind `e` to the `[]` invocation for this entry.
+
+        4.  Bind invocations in the entry value subpattern using target `e`.
+
+*   **Record**:
+
+    1.  For each field `f` in `p`:
+
+        1.  Let `f` be `i : (field, [])` where `field` is the corresponding
+            getter name for the field.
+
+        2.  Bind `e` to the field accessor for this field.
+
+        3.  Bind invocations in the field subpattern using target `e`.
+
+*   **Extractor**:
+
+    1.  For each field `f` in `p`:
+
+        1.  Resolve the getter name to a member.
+
+        2.  If it resolves to an instance method, then let `f` be `i : (field,
+            [])` where `field` is the name of the getter.
+
+        3.  Else it resolves to an extension method. Let `f` be `i : type(field,
```

Comment:
There's a subtlety here around what `type` is that should probably be specified.  I *think* (but am not sure), that what it should be is actually the name + type arguments of the extension declaration.  The resolved `on` type may not be sufficient, though maybe the fact that we can't use the explicit invocation syntax (`E(receiver).m`) here makes it sufficient? 
 An alternative would be the type of the receiver, but that may conflate more things than necessary.

We clearly need to make sure that things with different resolved `on` types are different.  Example:

```dart
extension E<T> on  List<T> {
  int get foo => (T == int) ? 3 : 4;
}

void test(Object x) {
  switch (x) {
    case List<int>(:var foo): ...
    case List<double>(:var foo):...
}
```

So it's not enough to use just `E` as the key.  Using `E<double>` and `E<int>` as the keys would work.

Using the type of the receiver is probably correct, because I think for a given name `foo`, two invocations of it with the same receiver type in the same import environment *must* always result in the same resolution.  This will be coarser than necessary though:

```dart
extension E on  num {
  int get foo => 3;
}

void test(Object x) {
  switch (x) {
    case int(:var foo): ...
    case double(:var foo):...
}
```

These two invocations are demonstrably equal, but if we use the receiver types as keys, they will be memoized separately.

Using the resolved `on` type would not be correct if we could somehow get an explicit invocation.  Consider this extension method declaration:

```dart
extension E<T> on  num {
  int get foo => (T == int) ? 3 : 4;
}
```

Clearly `E<int>(receiver).foo` and `E<double>(receiver).foo` will produce different results.  But I can't think offhand of any way to get this without using the explicit syntax, and I don't think we can use the explicit syntax here.

On balance, I *think* that what we want here is the extension name + type arguments as the key.  @lrhn and @eernstg might want to check my reasoning here though.

Reply:
I've already commented on this above. You pretty certainly need to remember the entire corresponding explicit extension invocation of `Ext<T>(x).foo`, which means the chosen extension declaration plus the inferred type arguments.

The type arguments are inferred from the static type of the receiver, but it's more precise to use the inferred type arguments. Take:
```dart
extension F<T> on Iterable<T> {
  (T, Type)? get firstAndTypeOrNull => this.isEmpty ? null : (this.first, T);
}
```
It behaves differently on `List<num>` and `List<int>`, so the inferred type must be included.
It resolves to the extension `F<int>` on both `List<int>` and `Iterable<int>`, so the entire receiver type is not important. 

The most precise distinguisher for behavior is the `F<int>` extension instantiation.


(Secondary note: It's slightly worrying that you *cannot* do an explicit extension instantiation.)

Reply:
I can't find where you commented above, and I can't tell whether you're agreeing with me or not.  I think you are?  That is, that it's sufficient to use the extension name + type arguments as the token?

Reply:
Ah, good catch! Fixed.